### ISSUE TITLE
date field show improper values

### ISSUE THREAD

**Issue by abuvanth** [getredash/redash#2074]

**Comment by arikfr** [comment]
There isn't much we can do with this issue without you providing more context and details.

**Comment by abuvanth** [comment]
display utc milli second instead of date

Acutal output
![acutal outdate](https://user-images.githubusercontent.com/20999104/32429106-27ac8718-c2ef-11e7-855b-ff62fa173c7a.png)

Expected output
![expected outdate](https://user-images.githubusercontent.com/20999104/32429108-28791314-c2ef-11e7-8916-3bce90ef4c38.png)

**Comment by abuvanth** [comment]
when i export data as csv it gives correct data.. but report gives milli second instead of date

thanks in advance.

**Comment by arikfr** [comment]
What version of Redash are you using?

**Comment by abuvanth** [comment]
version 3.0.0

**Comment by arikfr** [comment]
If you run the following query:

```
select null as t

union all

select now() as t
```

What do you get as results?

**Comment by abuvanth** [comment]
im using sql server so i use getdate() instead of now().

following is the ouput

 06/11/17 12:52

**Comment by abuvanth** [comment]
datatype of Dateout column is date

**Comment by abuvanth** [comment]
dateIn and dateout both are date datatype dateIn column show date but dateout coumn show millisecond

![dateinout](https://user-images.githubusercontent.com/20999104/32430084-992c1eb8-c2f3-11e7-96aa-cb8a3d6a39d8.png)

when i export data as csv, that show correct values

![csv](https://user-images.githubusercontent.com/20999104/32430095-af0d37bc-c2f3-11e7-8e99-82f516f1d743.png)

**Comment by arikfr** [comment]
Is the first row a null or an empty string? Can you limit the output to 2 rows and paste here the JSON format of the results?

**Comment by abuvanth** [comment]
you are correct dateIn field does not contain NULL values but Dateout columns contain null values

**Comment by abuvanth** [comment]
how can i fix this issue?

**Comment by abuvanth** [comment]
{
          "Owning_Railway": "SCR",
          "RetDate": null,
          "Workshop": null,
          "DateIN": "2017-07-20",
          "Depot_Base": "SC",
          "Trp_Code": "WRRMDAC",
          "RakeForm": null,
          "Built": null,
          "Mfg_Type": "ICF",
          "retdate1": null,
          "DateOut": null,
          "coachno": "99855",
          "RepType": "POH",
          "LPOHDate": "2016-02-01"
        },
        {
          "Owning_Railway": "SCR",
          "RetDate": "2019-11-01",
          "Workshop": null,
          "DateIN": "2016-03-26",
          "Depot_Base": "SC",
          "Trp_Code": "SPART",
          "RakeForm": "LOOSE",
          "Built": 1999,
          "Mfg_Type": "ICF",
          "retdate1": "01-11-2019",
          "DateOut": "2016-04-22",
          "coachno": "99911",
          "RepType": "POH",
          "LPOHDate": " - "
}

**Comment by arikfr** [comment]
The full query results data with the columns defintions please.

**Comment by abuvanth** [comment]
[query result.txt](https://github.com/getredash/redash/files/1445598/query.result.txt)

**Comment by arikfr** [comment]
The reason it happens is because the column type is marked as string and the first value is null. If you can get rid of the null or change the type of the column to DATETIME, it will work.

**Comment by arikfr** [comment]
I updated the issue description to contain more details.

Thank you for reporting.

**Comment by abuvanth** [comment]
thanks.

**Comment by tonyjiangh** [comment]
Is this one already resolved?
* the new table visulization editor allow you to decide display data type and format
* as @arikfr said, simply by defineing column type in SQL query, this problem could be easily avoid

```sql
select date('2016-04-09') as t
union all
select date(null) as t
```

![image](https://user-images.githubusercontent.com/16881036/35468578-6b275760-0364-11e8-80e1-59d5403a493a.png)

**Comment by arikfr** [comment]
Thanks for the follow up @tonyjiangh. The issue still exists, but as you mention we have now more options to avoid it.

**Comment by tonyjiangh** [comment]
I get it now, so I think your stance here is to tackle the issue at the conversion phase.

**Comment by arikfr** [comment]
> Or do we want to save the data as it is and tackle this issue at frontend?

Yes :)

Although while giving this more thought, I realized that now that the user can control how each column is rendered maybe we can simplify the logic around this. The result I would expect in such case, that this column gets rendered as is (as a `STRING`) and if the user wants they can choose to mark it as a date column.

**Comment by tonyjiangh** [comment]
That sounds great since I do recognize that you don't get to choose data type for many visualization before.
I think there are two ways to do this.

**Have one setting for all visualization under a query**, maybe use the default TABLE visualization setting as the global setting.

OR

**Each visulization get to choose it's own way of representation**.
Than we might need to work on:
* [Chart] set default representation for the field used for Group by
* [Pivot Table] set representation for fields used for row / column group
* [Map] group representation
* [Sankey/Sunburst] stage name representation
